grr/server/grr_response_server/flows/general/osquery.py

```diff
@@ -85,8 +200,51 @@ def Process(
 
     self._UpdateProgress(responses)
 
+    # Make Osquery results available before files are collected
     for response in responses:
       self.SendReply(response)
 
+    self._FileCollectionFromColumns(responses)
+
   def GetProgress(self) -> rdf_osquery.OsqueryProgress:
     return self.state.progress
+
+  def ReceiveFetchedFile(self,
+                         stat_entry,
+                         file_hash,
+                         request_data=None,
+                         is_duplicate=False) -> None:
+    self.SendReply(stat_entry)
+
+  def GetFilesArchiveMappings(
+      self,
+      flow_results: Iterator[rdf_flow_objects.FlowResult],
+  ) -> Iterator[flow_base.ClientPathArchiveMapping]:
+    def _GetUniqueSuffixFor(target_path_so_far: str) -> str:
+      times_used_before = self.state.path_to_count.get(target_path_so_far, 0)
+      self.state.path_to_count[target_path_so_far] = times_used_before + 1
+
+      if times_used_before > 0:
+        return f"-{times_used_before}"
+      else:
+        return ""
+
+    def GenerateTargetPath(pathspec: rdf_paths.PathSpec) -> str:
+      path_info = rdf_objects.PathInfo.FromPathSpec(pathspec)
+      client_string_path = "/".join(path_info.components)
+
+      target_path = f"osquery_collected_files/{client_string_path}"
+      target_path += _GetUniqueSuffixFor(target_path)
+      return target_path
```

Comment:
These utilities are used only once and declared here, making the code quite deeply nested. Sometimes refactoring logic into utility improves readability, here I feel like it only introduces unnecessary complexity.

If they make sense outside of this class, I would move them to stand alone methods. If they don't, I would just inline  the definition.

Reply:
Even though the logic is not massively complicated, I would like to keep at least one of those methods as a separate named block in the code, since it is not exactly trivial (e.g. we are maintaining state for the number of paths that we have seen).

I will try to inline `_GetUniqueSuffixFor` into `_GenerateTargetArchivePath`.

Reply:
I agree nesting is a bit controversial; I will move it out as a class method.

Reply:
Class methods are disencouraged by our internal style guidelines. They are okay in some cases, but I don't think this is one of them. Why no make it a top-level private method?

Reply:
Sorry I meant instance method.